pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPreprocessThrottler.java

```diff
@@ -0,0 +1,147 @@
+ * Code paths that do no need to rebuild the index or which don't happen on the server need not utilize this throttler.
+ */
+public class SegmentPreprocessThrottler implements PinotClusterConfigChangeListener {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentPreprocessThrottler.class);
+
+  /**
+   * _maxPreprocessConcurrency must be >= 0. To effectively disable throttling, this can be set to a very high value
+   */
+  private int _maxPreprocessConcurrency;
+  private boolean _relaxThrottling;
+  private final AdjustableSemaphore _semaphore;
+
+  /**
+   * @param maxPreprocessConcurrency configured preprocessing concurrency
+   * @param maxConcurrentPreprocessesBeforeServingQueries configured preprocessing concurrency before serving queries
+   * @param relaxThrottling whether to relax throttling prior to serving queries
+   */
+  public SegmentPreprocessThrottler(int maxPreprocessConcurrency, int maxConcurrentPreprocessesBeforeServingQueries,
+      boolean relaxThrottling) {
+    LOGGER.info("Initializing SegmentPreprocessThrottler, maxPreprocessConcurrency: {}, "
+            + "maxConcurrentPreprocessesBeforeServingQueries: {}, relaxThrottling: {}",
+        maxPreprocessConcurrency, maxConcurrentPreprocessesBeforeServingQueries, relaxThrottling);
+    Preconditions.checkArgument(maxPreprocessConcurrency > 0,
+        "Max preprocess parallelism must be >= 0, but found to be: " + maxPreprocessConcurrency);
+    Preconditions.checkArgument(maxConcurrentPreprocessesBeforeServingQueries > 0,
+        "Max preprocess parallelism before serving queries must be >= 0, but found to be: "
+            + maxConcurrentPreprocessesBeforeServingQueries);
+
+    _maxPreprocessConcurrency = maxPreprocessConcurrency;
+    _relaxThrottling = relaxThrottling;
+
+    // maxConcurrentPreprocessesBeforeServingQueries is only used prior to serving queries and once the server is
+    // ready to serve queries this is not used again. Thus, it is safe to only pick up this configuration during
+    // server startup. There is no need to allow updates to this via the ZK CLUSTER config handler
+    int relaxThrottlingThreshold = Math.max(_maxPreprocessConcurrency, maxConcurrentPreprocessesBeforeServingQueries);
+    int preprocessConcurrency = _maxPreprocessConcurrency;
+    if (relaxThrottling) {
+      preprocessConcurrency = relaxThrottlingThreshold;
+      LOGGER.info("Relax throttling enabled, setting preprocess concurrency to: {}", preprocessConcurrency);
+    }
+    _semaphore = new AdjustableSemaphore(preprocessConcurrency, true);
+    LOGGER.info("Created semaphore with available permits: {}", _semaphore.availablePermits());
+  }
+
+  public synchronized void resetThrottling() {
+    LOGGER.info("Reset throttling threshold for segment preprocess concurrency, available permits: {}",
+        _semaphore.availablePermits());
+    _relaxThrottling = false;
+    _semaphore.setPermits(_maxPreprocessConcurrency);
```

Comment:
should `setPermits()` method in AdjustableSemaphore be `synchronized`? as it may read-check-modify the sema and _totalPermits, potentially from different threads. 

But I see setPermits() happens to be called in methods with `synchronized` so far. If it's required to use within synchronized method, then perhaps better comment it out for AdjustableSemaphore class or just make it synchronized (cc @yashmayya fyi)

Reply:
I think @yashmayya is working on some synchronization improvement to make '_setPermits` volatile in another PR. @yashmayya perhaps you can add a comment on expected usage of `_setPermits` as part of your PR?

Reply:
@klsince @somandal the `AdjustableSemaphore` is currently only being used in the `MultiStageQueryThrottler` where `setPermits` can't be called concurrently from multiple threads. We made `_totalPermits` volatile because of the new public method `getTotalPermits`. Essentially concurrent reads during writes are allowed but not concurrent writes. If we want to support concurrent writes as well we should definitely make the `setPermits` method `synchronized`.

Reply:
@yashmayya @klsince I've made the AdjustableSemaphore::setPermits() function synchronized to allow for concurrent writers. Please take a look when you get a chance. thanks!